File: js/tests/helpers/mocks.js
/**
 * Mock utilities for unit tests
 */

/**
 * Create a mock Playwright page object
 */
export function createMockPlaywrightPage(options = {}) {
  const {
    url = 'https://example.com',
    elements = {},
    evaluateResult = null,
  } = options;

  const eventListeners = new Map();

  const locatorMock = (selector) => {
    const elementData = elements[selector] || {
      count: 1,
      visible: true,
      enabled: true,
      textContent: 'Mock text',
      value: '',
      className: 'mock-class',
    };

    return {
      count: async () => elementData.count,
      first() {
        return this;
      },
      nth(i) {
        return this;
      },
      last() {
        return this;
      },
      click: async (opts = {}) => {},
      fill: async (text) => {
        elementData.value = text;
      },
      type: async (text) => {
        elementData.value = text;
      },
      focus: async () => {},
      textContent: async () => elementData.textContent,
      inputValue: async () => elementData.value,
      getAttribute: async (name) => elementData[name] || null,
      isVisible: async () => elementData.visible,
      waitFor: async (opts = {}) => {
        if (!elementData.visible && opts.state === 'visible') {
          throw new Error('Element not visible');
        }
      },
      evaluate: async (fn, arg) => {
        const mockEl = {
          tagName: 'DIV',
          textContent: elementData.textContent,
          value: elementData.value,
          className: elementData.className,
          disabled: !elementData.enabled,
          checked: elementData.checked || false,
          isConnected: true,
          offsetWidth: elementData.visible ? 100 : 0,
          offsetHeight: elementData.visible ? 50 : 0,
          hasAttribute: (attr) => !!elementData[attr],
          getAttribute: (attr) => elementData[attr] || null,
          classList: {
            contains: (cls) => elementData.className?.includes(cls) || false,
          },
          getBoundingClientRect: () => ({
            top: 100,
            bottom: 150,
            left: 10,
            right: 110,
            width: 100,
            height: 50,
          }),
          scrollIntoView: () => {},
          dispatchEvent: () => {},
        };
        return fn(mockEl, arg);
      },
    };
  };

  return {
    _isPlaywrightPage: true,
    url: () => url,
    goto: async (targetUrl, opts = {}) => {},
    waitForNavigation: async (opts = {}) => {},
    waitForSelector: async (sel, opts = {}) => locatorMock(sel),
    $: async (sel) => locatorMock(sel),
    $$: async (sel) => {
      const count = elements[sel]?.count || 1;
      return Array(count).fill(locatorMock(sel));
    },
    $eval: async (sel, fn, ...args) => {
      const loc = locatorMock(sel);
      return loc.evaluate(fn, ...args);
    },
    $$eval: async (sel, fn, ...args) => {
      const count = elements[sel]?.count || 1;
      const locs = Array(count).fill(locatorMock(sel));
      return locs.map((l) => l.evaluate(fn, ...args));
    },
    locator: locatorMock,
    // Playwright's page.evaluate() only accepts a single argument (not spread)
    // This is the key difference from Puppeteer
    evaluate: async (fn, arg) => {
      if (evaluateResult !== null) {
        return evaluateResult;
      }
      try {
        // Playwright passes exactly one argument (can be object/array)
        return fn(arg);
      } catch {
        return fn;
      }
    },
    mainFrame: () => ({
      url: () => url,
    }),
    context: () => ({}),
    bringToFront: async () => {},
    on: (event, handler) => {
      if (!eventListeners.has(event)) {
        eventListeners.set(event, []);
      }
      eventListeners.get(event).push(handler);
    },
    off: (event, handler) => {
      const handlers = eventListeners.get(event);
      if (handlers) {
        const idx = handlers.indexOf(handler);
        if (idx !== -1) {
          handlers.splice(idx, 1);
        }
      }
    },
    emit: (event, data) => {
      const handlers = eventListeners.get(event);
      if (handlers) {
        handlers.forEach((h) => h(data));
      }
    },
    click: async (sel, opts = {}) => {},
    type: async (sel, text, opts = {}) => {},
    keyboard: {
      type: async (text) => {},
    },
  };
}

/**
 * Create a mock Puppeteer page object
 */
export function createMockPuppeteerPage(options = {}) {
  const {
    url = 'https://example.com',
    elements = {},
    evaluateResult = null,
  } = options;

  const eventListeners = new Map();

  const elementMock = (selector) => {
    const elementData = elements[selector] || {
      count: 1,
      visible: true,
      enabled: true,
      textContent: 'Mock text',
      value: '',
      className: 'mock-class',
    };

    return {
      click: async (opts = {}) => {},
      focus: async () => {},
      type: async (text) => {
        elementData.value = text;
      },
      evaluate: async (fn, ...args) =>
        page.evaluate(fn, elementMock(selector), ...args),
    };
  };

  const page = {
    _isPuppeteerPage: true,
    url: () => url,
    goto: async (targetUrl, opts = {}) => {},
    waitForNavigation: async (opts = {}) => {},
    waitForSelector: async (sel, opts = {}) => elementMock(sel),
    $: async (sel) => {
      const elementData = elements[sel];
      if (elementData?.count === 0) {
        return null;
      }
      return elementMock(sel);
    },
    $$: async (sel) => {
      const count = elements[sel]?.count || 1;
      return Array(count)
        .fill(null)
        .map(() => elementMock(sel));
    },
    $eval: async (sel, fn, ...args) => {
      const elementData = elements[sel] || {};
      const mockEl = {
        tagName: 'DIV',
        textContent: elementData.textContent || 'Mock text',
        value: elementData.value || '',
        className: elementData.className || 'mock-class',
        disabled: !elementData.enabled,
        checked: elementData.checked || false,
        isConnected: true,
        offsetWidth: elementData.visible !== false ? 100 : 0,
        offsetHeight: elementData.visible !== false ? 50 : 0,
        hasAttribute: (attr) => !!elementData[attr],
        getAttribute: (attr) => elementData[attr] || null,
        classList: {
          contains: (cls) => elementData.className?.includes(cls) || false,
        },
        getBoundingClientRect: () => ({
          top: 100,
          bottom: 150,
          left: 10,
          right: 110,
          width: 100,
          height: 50,
        }),
        scrollIntoView: () => {},
        dispatchEvent: () => {},
      };
      return fn(mockEl, ...args);
    },
    $$eval: async (sel, fn, ...args) => {
      const count = elements[sel]?.count || 1;
      return fn(Array(count).fill({}), ...args);
    },
    evaluate: async (fn, ...args) => {
      if (evaluateResult !== null) {
        return evaluateResult;
      }
      try {
        // For element evaluations, first arg might be an element mock
        if (args[0] && typeof args[0] === 'object' && args[0].click) {
          const mockEl = {
            tagName: 'DIV',
            textContent: 'Mock text',
            value: '',
            className: 'mock-class',
            disabled: false,
            checked: false,
            isConnected: true,
            offsetWidth: 100,
            offsetHeight: 50,
            hasAttribute: () => false,
            getAttribute: () => null,
            classList: { contains: () => false },
            getBoundingClientRect: () => ({
              top: 100,
              bottom: 150,
              left: 10,
              right: 110,
              width: 100,
              height: 50,
            }),
            scrollIntoView: () => {},
            dispatchEvent: () => {},
          };
          return fn(mockEl, ...args.slice(1));
        }
        return fn(...args);
      } catch {
        return fn;
      }
    },
    mainFrame: () => ({
      url: () => url,
    }),
    bringToFront: async () => {},
    on: (event, handler) => {
      if (!eventListeners.has(event)) {
        eventListeners.set(event, []);
      }
      eventListeners.get(event).push(handler);
    },
    off: (event, handler) => {
      const handlers = eventListeners.get(event);
      if (handlers) {
        const idx = handlers.indexOf(handler);
        if (idx !== -1) {
          handlers.splice(idx, 1);
        }
      }
    },
    emit: (event, data) => {
      const handlers = eventListeners.get(event);
      if (handlers) {
        handlers.forEach((h) => h(data));
      }
    },
    click: async (sel, opts = {}) => {},
    type: async (sel, text, opts = {}) => {},
    keyboard: {
      type: async (text) => {},
    },
  };

  return page;
}

/**
 * Create a mock logger
 */
export function createMockLogger(options = {}) {
  const { collectLogs = false } = options;
  const logs = [];

  return {
    debug: (fn) => {
      if (collectLogs) {
        logs.push({
          level: 'debug',
          message: typeof fn === 'function' ? fn() : fn,
        });
      }
    },
    info: (fn) => {
      if (collectLogs) {
        logs.push({
          level: 'info',
          message: typeof fn === 'function' ? fn() : fn,
        });
      }
    },
    warn: (fn) => {
      if (collectLogs) {
        logs.push({
          level: 'warn',
          message: typeof fn === 'function' ? fn() : fn,
        });
      }
    },
    error: (fn) => {
      if (collectLogs) {
        logs.push({
          level: 'error',
          message: typeof fn === 'function' ? fn() : fn,
        });
      }
    },
    getLogs: () => logs,
    clear: () => {
      logs.length = 0;
    },
  };
}

/**
 * Create a mock network tracker
 */
export function createMockNetworkTracker(options = {}) {
  const { initialPendingCount = 0, waitForIdleResult = true } = options;

  let pendingCount = initialPendingCount;
  const listeners = {
    onRequestStart: [],
    onRequestEnd: [],
    onNetworkIdle: [],
  };

  return {
    startTracking: () => {},
    stopTracking: () => {},
    waitForNetworkIdle: async (opts = {}) => waitForIdleResult,
    getPendingCount: () => pendingCount,
    getPendingUrls: () => [],
    reset: () => {
      pendingCount = 0;
    },
    on: (event, callback) => {
      if (listeners[event]) {
        listeners[event].push(callback);
      }
    },
    off: (event, callback) => {
      if (listeners[event]) {
        const idx = listeners[event].indexOf(callback);
        if (idx !== -1) {
          listeners[event].splice(idx, 1);
        }
      }
    },
    setPendingCount: (count) => {
      pendingCount = count;
    },
  };
}

/**
 * Create a mock navigation manager
 */
export function createMockNavigationManager(options = {}) {
  const {
    currentUrl = 'https://example.com',
    isNavigating = false,
    shouldAbortValue = false,
  } = options;

  let url = currentUrl;
  let navigating = isNavigating;
  const sessionId = 1;
  let abortController = new AbortController();
  const listeners = {
    onNavigationStart: [],
    onNavigationComplete: [],
    onBeforeNavigate: [],
    onUrlChange: [],
    onPageReady: [],
  };

  return {
    navigate: async (opts = {}) => {
      url = opts.url || url;
      return true;
    },
    waitForNavigation: async (opts = {}) => true,
    waitForPageReady: async (opts = {}) => true,
    isNavigating: () => navigating,
    getCurrentUrl: () => url,
    getSessionId: () => sessionId,
    getAbortSignal: () => abortController.signal,
    shouldAbort: () => shouldAbortValue,
    onSessionCleanup: (callback) => {},
    on: (event, callback) => {
      if (listeners[event]) {
        listeners[event].push(callback);
      }
    },
    off: (event, callback) => {
      if (listeners[event]) {
        const idx = listeners[event].indexOf(callback);
        if (idx !== -1) {
          listeners[event].splice(idx, 1);
        }
      }
    },
    startListening: () => {},
    stopListening: () => {},
    configure: (config) => {},
    setUrl: (newUrl) => {
      url = newUrl;
    },
    setNavigating: (val) => {
      navigating = val;
    },
    triggerEvent: (event, data) => {
      if (listeners[event]) {
        listeners[event].forEach((cb) => cb(data));
      }
    },
    abort: () => {
      abortController.abort();
    },
    resetAbort: () => {
      abortController = new AbortController();
    },
  };
}

/**
 * Create navigation error for testing
 */
export function createNavigationError(
  message = 'Execution context was destroyed'
) {
  const error = new Error(message);
  error.name = 'NavigationError';
  return error;
}

/**
 * Simple assertion helpers that work with Node.js test runner
 */
export const assert = {
  equal: (actual, expected, message = '') => {
    if (actual !== expected) {
      throw new Error(`${message}: Expected ${expected}, got ${actual}`);
    }
  },
  deepEqual: (actual, expected, message = '') => {
    if (JSON.stringify(actual) !== JSON.stringify(expected)) {
      throw new Error(`${message}: Objects not equal`);
    }
  },
  ok: (value, message = '') => {
    if (!value) {
      throw new Error(`${message}: Expected truthy value, got ${value}`);
    }
  },
  notOk: (value, message = '') => {
    if (value) {
      throw new Error(`${message}: Expected falsy value, got ${value}`);
    }
  },
  throws: async (fn, expectedError, message = '') => {
    try {
      await fn();
      throw new Error(`${message}: Expected function to throw`);
    } catch (error) {
      if (expectedError && !error.message.includes(expectedError)) {
        throw new Error(
          `${message}: Expected error message to include "${expectedError}", got "${error.message}"`
        );
      }
    }
  },
  doesNotThrow: async (fn, message = '') => {
    try {
      await fn();
    } catch (error) {
      throw new Error(
        `${message}: Expected function not to throw, but got: ${error.message}`
      );
    }
  },
};

File: python/tests/helpers/mocks.py
"""Mock utilities for unit tests."""

from __future__ import annotations

import asyncio
import contextlib
from dataclasses import dataclass
from typing import Any, Callable
from unittest.mock import AsyncMock, MagicMock


@dataclass
class MockElementData:
    """Mock element data."""

    count: int = 1
    visible: bool = True
    enabled: bool = True
    text_content: str = "Mock text"
    value: str = ""
    class_name: str = "mock-class"
    checked: bool = False


def create_mock_playwright_page(
    url: str = "https://example.com",
    elements: dict[str, MockElementData] | None = None,
    evaluate_result: Any = None,
) -> MagicMock:
    """Create a mock Playwright page object.

    Args:
        url: Current page URL
        elements: Mock element data by selector
        evaluate_result: Result to return from evaluate

    Returns:
        Mock Playwright page
    """
    elements = elements or {}
    event_listeners: dict[str, list[Callable]] = {}

    def get_element_data(selector: str) -> MockElementData:
        return elements.get(selector, MockElementData())

    def create_locator(selector: str) -> MagicMock:
        element_data = get_element_data(selector)
        locator = MagicMock()
        locator.count = AsyncMock(return_value=element_data.count)
        locator.first.return_value = locator
        locator.nth.return_value = locator
        locator.last.return_value = locator
        locator.click = AsyncMock()
        locator.fill = AsyncMock()
        locator.type = AsyncMock()
        locator.focus = AsyncMock()
        locator.text_content = AsyncMock(return_value=element_data.text_content)
        locator.input_value = AsyncMock(return_value=element_data.value)
        locator.get_attribute = AsyncMock(return_value=None)
        locator.is_visible = AsyncMock(return_value=element_data.visible)
        locator.wait_for = AsyncMock()

        async def mock_evaluate(fn: Callable, arg: Any = None) -> Any:
            mock_el = MagicMock()
            mock_el.tagName = "DIV"
            mock_el.textContent = element_data.text_content
            mock_el.value = element_data.value
            mock_el.className = element_data.class_name
            mock_el.disabled = not element_data.enabled
            mock_el.checked = element_data.checked
            mock_el.isConnected = True
            mock_el.offsetWidth = 100 if element_data.visible else 0
            mock_el.offsetHeight = 50 if element_data.visible else 0
            return fn(mock_el, arg)

        locator.evaluate = mock_evaluate
        return locator

    page = MagicMock()
    page._is_playwright_page = True
    page.url.return_value = url
    page.goto = AsyncMock()
    page.wait_for_navigation = AsyncMock()
    page.wait_for_selector = AsyncMock(side_effect=lambda s, **_kw: create_locator(s))
    page.query_selector = AsyncMock(side_effect=lambda s: create_locator(s))
    page.query_selector_all = AsyncMock(
        side_effect=lambda s: [create_locator(s)] * get_element_data(s).count
    )
    page.locator = create_locator

    async def mock_evaluate(fn: Callable, arg: Any = None) -> Any:
        if evaluate_result is not None:
            return evaluate_result
        try:
            return fn(arg)
        except Exception:
            return fn

    page.evaluate = mock_evaluate
    page.main_frame.return_value.url.return_value = url
    page.context.return_value = MagicMock()
    page.bring_to_front = AsyncMock()

    def on_handler(event: str, handler: Callable) -> None:
        if event not in event_listeners:
            event_listeners[event] = []
        event_listeners[event].append(handler)

    def off_handler(event: str, handler: Callable) -> None:
        if event in event_listeners:
            with contextlib.suppress(ValueError):
                event_listeners[event].remove(handler)

    def emit(event: str, data: Any = None) -> None:
        if event in event_listeners:
            for h in event_listeners[event]:
                h(data)

    page.on = on_handler
    page.off = off_handler
    page.emit = emit
    page.click = AsyncMock()
    page.type = AsyncMock()
    page.keyboard = MagicMock()
    page.keyboard.type = AsyncMock()

    return page


def create_mock_selenium_driver(
    url: str = "https://example.com",
    elements: dict[str, MockElementData] | None = None,
) -> MagicMock:
    """Create a mock Selenium WebDriver.

    Args:
        url: Current page URL
        elements: Mock element data by selector

    Returns:
        Mock Selenium driver
    """
    elements = elements or {}

    def get_element_data(selector: str) -> MockElementData:
        return elements.get(selector, MockElementData())

    def create_element(selector: str) -> MagicMock:
        element_data = get_element_data(selector)
        element = MagicMock()
        element.click = MagicMock()
        element.send_keys = MagicMock()
        element.clear = MagicMock()
        element.text = element_data.text_content
        element.get_attribute = MagicMock(return_value=element_data.value)
        element.is_displayed = MagicMock(return_value=element_data.visible)
        element.is_enabled = MagicMock(return_value=element_data.enabled)
        element.location = {"x": 10, "y": 100}
        element.size = {"width": 100, "height": 50}
        return element

    driver = MagicMock()
    driver._is_selenium_driver = True
    driver.current_url = url
    driver.get = MagicMock()
    driver.find_element = MagicMock(side_effect=lambda _by, val: create_element(val))
    driver.find_elements = MagicMock(
        side_effect=lambda _by, val: [create_element(val)] * get_element_data(val).count
    )
    driver.execute_script = MagicMock(return_value=None)
    driver.execute_async_script = MagicMock(return_value=None)

    return driver


def create_mock_logger(collect_logs: bool = False) -> MagicMock:
    """Create a mock logger.

    Args:
        collect_logs: Whether to collect log entries

    Returns:
        Mock logger
    """
    logs: list[dict[str, Any]] = []

    def make_log_fn(level: str) -> Callable:
        def log_fn(fn_or_msg: Callable | str) -> None:
            if collect_logs:
                msg = fn_or_msg() if callable(fn_or_msg) else fn_or_msg
                logs.append({"level": level, "message": msg})

        return log_fn

    logger = MagicMock()
    logger.debug = make_log_fn("debug")
    logger.info = make_log_fn("info")
    logger.warn = make_log_fn("warn")
    logger.error = make_log_fn("error")
    logger.get_logs = lambda: logs
    logger.clear = lambda: logs.clear()

    return logger


def create_mock_network_tracker(
    initial_pending_count: int = 0,
    wait_for_idle_result: bool = True,
) -> MagicMock:
    """Create a mock network tracker.

    Args:
        initial_pending_count: Initial number of pending requests
        wait_for_idle_result: Result for wait_for_network_idle

    Returns:
        Mock network tracker
    """
    pending_count = initial_pending_count
    listeners: dict[str, list[Callable]] = {
        "on_request_start": [],
        "on_request_end": [],
        "on_network_idle": [],
    }

    tracker = MagicMock()
    tracker.start_tracking = MagicMock()
    tracker.stop_tracking = MagicMock()
    tracker.wait_for_network_idle = AsyncMock(return_value=wait_for_idle_result)
    tracker.get_pending_count = MagicMock(return_value=pending_count)
    tracker.get_pending_urls = MagicMock(return_value=[])
    tracker.reset = MagicMock()

    def on_handler(event: str, callback: Callable) -> None:
        if event in listeners:
            listeners[event].append(callback)

    def off_handler(event: str, callback: Callable) -> None:
        if event in listeners:
            with contextlib.suppress(ValueError):
                listeners[event].remove(callback)

    tracker.on = on_handler
    tracker.off = off_handler

    return tracker


def create_mock_navigation_manager(
    current_url: str = "https://example.com",
    is_navigating: bool = False,
    should_abort_value: bool = False,
) -> MagicMock:
    """Create a mock navigation manager.

    Args:
        current_url: Current URL
        is_navigating: Whether currently navigating
        should_abort_value: Value for should_abort

    Returns:
        Mock navigation manager
    """
    url = current_url
    navigating = is_navigating
    session_id = 1
    abort_signal = asyncio.Event()

    listeners: dict[str, list[Callable]] = {
        "on_navigation_start": [],
        "on_navigation_complete": [],
        "on_before_navigate": [],
        "on_url_change": [],
        "on_page_ready": [],
    }

    manager = MagicMock()
    manager.navigate = AsyncMock(return_value=True)
    manager.wait_for_navigation = AsyncMock(return_value=True)
    manager.wait_for_page_ready = AsyncMock(return_value=True)
    manager.is_navigating = MagicMock(return_value=navigating)
    manager.get_current_url = MagicMock(return_value=url)
    manager.get_session_id = MagicMock(return_value=session_id)
    manager.get_abort_signal = MagicMock(return_value=abort_signal)
    manager.should_abort = MagicMock(return_value=should_abort_value)
    manager.start_listening = MagicMock()
    manager.stop_listening = MagicMock()

    def on_handler(event: str, callback: Callable) -> None:
        if event in listeners:
            listeners[event].append(callback)

    def off_handler(event: str, callback: Callable) -> None:
        if event in listeners:
            with contextlib.suppress(ValueError):
                listeners[event].remove(callback)

    manager.on = on_handler
    manager.off = off_handler

    return manager


def create_navigation_error(
    message: str = "Execution context was destroyed",
) -> Exception:
    """Create a navigation error for testing.

    Args:
        message: Error message

    Returns:
        Navigation error
    """
    error = Exception(message)
    error.name = "NavigationError"
    return error
